### PR TITLE
[SPARK-43156][SQL] Fix `COUNT(*) is null` bug in correlated scalar subquery

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
Example query:
```sql

create or replace temp view t0 (a, b)
as values
    (1, 1.0),
    (2, 2.0);
create or replace temp view t1 (c, d)
as values
    (2, 3.0); 

spark.sql("select *, (select (count(1)) is null from t1 where t0.a = t1.c) from t0").collect()
res6: Array[org.apache.spark.sql.Row] = Array([1,1.0,null], [2,2.0,false])  
```
In this subquery, count(1) always evaluates to a non-null integer value, so count(1) is null is always false. The correct evaluation of the subquery is always false.

We incorrectly evaluate it to null for empty groups. The reason is that NullPropagation rewrites `Aggregate [c] [isnull(count(1))]` to `Aggregate [c] [false]`, this rewrite would be correct normally, but in the context of a scalar subquery it breaks our count bug handling in RewriteCorrelatedScalarSubquery.constructLeftJoins . By the time we get there, the query appears to not have the count bug - it looks the same as if the original query had a subquery with `select any_value(false) from r...`, and that case is not subject to the count bug.

Postgres comparison show correct always-false result: http://sqlfiddle.com/#!17/67822/5

Solution:

The reason are `NullPropagation` triggers the bug for `(count(1)) is null`, but the root cause is the wrong handling of literals when dealing with the count bug. The bug can be triggered when the scalar subquery has a global aggregate with a constant, which can be produced by NullPropagation or by the user query directly. We should make sure `resultWithZeroTups` return real value when use scalar subquery and return value are `Literal`. 

### Why are the changes needed?

Fix `COUNT(*) is null bug` in correlated scalar subquery

### Does this PR introduce _any_ user-facing change?

Yes, some sql query result will be changed.

### How was this patch tested?

Add new tests

### CODE REVIEW

**Comment by Hisoka-X** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 1]
```diff
@@ -459,7 +459,8 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
       case alias @ Alias(_: AttributeReference, _) =>
         (alias.exprId, Literal.create(null, alias.dataType))
       case alias @ Alias(l: Literal, _) =>
-        (alias.exprId, l.copy(value = null))
+        // SPARK-43156: copy value to get the correct value when extracting in constructLeftJoins
+        (alias.exprId, l.copy(value = l.value))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:463]: what does `l.copy(value = l.value)` do? seems nothing?

Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala]: My fault, just want to create new `Literal` instance with seem value.

**Review by jchen5 — COMMENTED** [review]
> We can judge whether the column returned by subquery is foldable (already handle by NullPropagation). If it is, it means that the result of this value has no substantial relationship with the data, and the presence or absence of data will not affect this column. So in this case, this column can be extracted from the JOIN to ensure that this value can be obtained regardless of whether the data JOIN is successful or not.

Hmm, I don't think this logic is correct because of the issue with subquery aggregates being a bit special that I describe in the issue description: most of the time the expression `COUNT(*) is null` can indeed be reduced to `false`, but when they are the result expression of a subquery with empty input they are actually different - `(select COUNT(*) is null from empty_tbl)` evaluates false, but `(select any_value(false) from empty_tbl)` evaluates null. This is because on empty input, most aggregates return NULL, so the false here doesn't affect the output - but COUNT is an exception and does evaluate to 0.

I think your fix will fix the `COUNT(*) is null` case, but break the `false` case, e.g. in a query like `select *, (select any_value(false) as result from t1 where t0.a = t1.c) from t0`) Can you confirm this?

I actually started working on a fix at the same time too, my approach is different and I think is able to avoid that problem by computing the resultWithZeroTups at the decorrelation stage, before rules like NullPropagation and ConstantFolding have run.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 2]
```diff
@@ -599,10 +600,32 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
...
+            case _ =>
+              false
+          }
+        }
+
         if (resultWithZeroTups.isEmpty) {
           // CASE 1: Subquery guaranteed not to have the COUNT bug because it evaluates to NULL
           // with zero tuples.
           planWithoutCountBug
+        } else if (resultFoldable) {
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:623]: Already merged before create this PR. 

> if the subquery is something like `select false from ... group by c` then it will still actually return null on empty inputs.

That's why we should judge whether the column returned by subquery is foldable. If the result depends on input, the `resultFoldable` will be `false`.  It will skip the logic I created.

> I think your fix will fix the COUNT(*) is null case, but break the false case, e.g. in a query like select *, (select any_value(false) as result from t1 where t0.a = t1.c) from t0) Can you confirm this?

Is this answer right?
image

**Comment by cloud-fan** [comment]
> I think your fix will fix the COUNT(*) is null case, but break the false case

Hi @jchen5 , can you help add a test for it in a new PR? The current PR passes all tests which means our test coverage is not good.

**Comment by Hisoka-X** [comment]
> > I think your fix will fix the COUNT(*) is null case, but break the false case
> 
> Hi @jchen5 , can you help add a test for it in a new PR? The current PR passes all tests which means our test coverage is not good.

If you don't mind, I can also add the required test cases to my PR. However, after I merged #40811, it seems that the current use case has been able to prove the correctness of this PR.😂

**Comment by cloud-fan** [comment]
@Hisoka-X do you mean the bug is already fixed now?

**Comment by Hisoka-X** [comment]
> @Hisoka-X do you mean the bug is already fixed now?

Depends on what is correct results of `select *, (select any_value(false) as result from t1 where t0.a = t1.c) from t0)` ?

> if the subquery is something like select false from ... group by c then it will still actually return null on empty inputs.

The current state is as described by @jchen5  . `select false from ... group by c` will result null on empty inputs, not return `false`. I think this is what @jchen5 worry about. You can check by https://github.com/apache/spark/pull/40865#discussion_r1173210161 .

If you think `select false from ... group by c` in subquery should return `false` on empty inputs, this PR have bug now. If you think it should return `null` on empty inputs, this PR are correct.

**Comment by cloud-fan** [comment]
Before we make a decision, let's try the query with postgres, oracle, db2, etc.

**Comment by Hisoka-X** [comment]
> Before we make a decision, let's try the query with postgres, oracle, db2, etc.

Ok. By the way, I check with postgres, it return `null`
image

File: sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out [hunk 3]
```diff
@@ -106,14 +106,14 @@ from l
 -- !query schema
 struct<a:int,b:decimal(2,1),scalarsubquery(a):boolean>
 -- !query output
-1	2.0	NULL
```
Comment (allisonwang-db) [inline sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out:109]: These results are incorrect. They should be `NULL` instead of `false`.

Comment (Hisoka-X) [reply sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out:109]: Thanks for remind. I fixed this bug and add test to cover it. I will update `sql.out` later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (jchen5) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala]: That answer does look right, let me debug this code to figure out why it's right and if I can find a counterexample.

Your check needs to go after the check immediately below to be correct (that's the cause of the test results change Allison commented).

**Comment by jchen5** [comment]
> Depends on what is correct results of select *, (select any_value(false) as result from t1 where t0.a = t1.c) from t0) ?

Yes, this should return null on empty data.

I will try to find another test case to check the potential issue I mentioned. Maybe something like `select *, (select false as result from t1 where t0.a = t1.c) from t0 limit 1)`(correct answer for that is also null).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala]: Yes, I add `mayHaveCountBug.getOrElse(false)` into check part.

File: sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out [hunk 4]
```diff
@@ -86,14 +86,14 @@ from l
...
-1	2.0	NULL
+1	2.0	false
+1	2.0	false
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out:90]: @jchen5 was the old result wrong?

Comment (jchen5) [reply sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out:90]: Yes, old result here was wrong, new result is correct.

@Hisoka-X There's a comment about it in the input file, please update it to remove the "but this case is wrong due to bug SPARK-43156" part

Comment (Hisoka-X) [reply sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out:90]: Done

**Comment by jchen5** [comment]
I checked the case of `any_value(false)` in a debugger and it works because resultWithZeroTups is NULL there, so that explains why it works - because there's an aggregation value around false, and not just constant false.

This fix seems to work with the cases I've thought of (results match SQL standard semantics and postgres).

I think this PR is mergeable. I'll plan to add some more test coverage for related cases in a future PR.

**Review by jchen5 — APPROVED** [review]

**Review by jchen5 — COMMENTED** [review]
To help mitigate issues from the change in query results, we can add this logic under the flag SQLConf.DECORRELATE_SUBQUERY_LEGACY_INCORRECT_COUNT_HANDLING_ENABLED (if false, use the new logic, if true, use old logic).

**Comment by Hisoka-X** [comment]
kindly ping @cloud-fan . All CI passed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 5]
```diff
@@ -459,7 +459,8 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
       case alias @ Alias(_: AttributeReference, _) =>
         (alias.exprId, Literal.create(null, alias.dataType))
       case alias @ Alias(l: Literal, _) =>
-        (alias.exprId, l.copy(value = null))
+        // SPARK-43156: copy value to get the correct value when extracting in constructLeftJoins
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:462]: `Literal` is immutable, why a copy is needed?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 6]
```diff
@@ -599,10 +600,33 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
...
           // CASE 1: Subquery guaranteed not to have the COUNT bug because it evaluates to NULL
           // with zero tuples.
           planWithoutCountBug
+        } else if (mayHaveCountBug.getOrElse(false) && resultFoldable &&
+          !conf.getConf(SQLConf.DECORRELATE_SUBQUERY_LEGACY_INCORRECT_COUNT_HANDLING_ENABLED)) {
+          val alias = Alias(resultWithZeroTups.get, origOutput.name)()
+          subqueryAttrMapping += (origOutput -> alias.toAttribute)
+          Project(
+            currentChild.output :+ alias,
+            Join(currentChild, query, LeftOuter, conditions.reduceOption(And), joinHint))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:629]: @allisonwang-db does the new decorrelation framework need the same fix?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 5 above)
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala]: My mistake. Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 7]
```diff
@@ -599,10 +600,33 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
         if (Utils.isTesting) {
           assert(mayHaveCountBug.isDefined)
         }
+
+        def queryOutputFoldable(list: Seq[NamedExpression]): Boolean = {
+          trimAliases(list.filter(p => p.exprId.equals(query.output.head.exprId)).head).foldable
+        }
+
+        lazy val resultFoldable = {
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:613]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 6 above)
Comment (jchen5) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:634]: This change works for the new decorrelation framework (and the old one too). This code happens after / on top of the DecorrelateInnerQuery framework.

Comment (jchen5) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:634]: Quick PR to add some test coverage for it: https://github.com/apache/spark/pull/40946

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 8]
```diff
@@ -459,7 +459,7 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
       case alias @ Alias(_: AttributeReference, _) =>
         (alias.exprId, Literal.create(null, alias.dataType))
       case alias @ Alias(l: Literal, _) =>
-        (alias.exprId, l.copy(value = null))
+        (alias.exprId, l)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:462]: I think this one-line change is all we need. Count bug does not apply to the count function only, but all expressions in Aggregate that return non-null value for empty input. This includes all constant values.

Assume there are two table `t1(c1 int)`, `t2(c2 int)`. `t2` has two rows (1, 2), `t3` has two rows (1,3)

```
scala> sql("select (select f from (select false as f, max(c1) from t1 where c1 = c2)) from t2").show
+------------------+
|scalarsubquery(c2)|
+------------------+
|             false|
|              NULL|
+------------------+
```
The result wrong and it should be false, false

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:462]: cc @jchen5

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:462]: We can optimize the code a bit, to build simpler plan if `resultWithZeroTups.get` is a `Literal`

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala [hunk 9]
```diff
@@ -1928,6 +1928,15 @@ class SubquerySuite extends QueryTest
     checkAnswer(df, Nil)
   }
 
+  test("SPARK-43156: scalar subquery with Literal result like `COUNT(1) is null`") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala:1931]: this is no longer needed as we have golden file test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 8 above)
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:462]: > We can optimize the code a bit, to build simpler plan if resultWithZeroTups.get is a Literal

Can't, in `select *, (select count(1) from t2 where t1.c1 = t2.c1) from t1`. The resultWithZeroTups.get will return `Literal(0)`. That's why we need use `resultFoldable`.

File: sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala (hunk 9 above)
Comment (Hisoka-X) [reply sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala]: Removed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 8 above)
Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:462]: > I think this one-line change is all we need. Count bug does not apply to the count function only, but all expressions in Aggregate that return non-null value for empty input. This includes all constant values.
> 
> Assume there are two table `t1(c1 int)`, `t2(c2 int)`. `t1` has two rows (1, 2), `t2` has two rows (1,3)
> 
> ```
> scala> sql("select (select f from (select false as f, max(c1) from t1 where c1 = c2)) from t2").show
> +------------------+
> |scalarsubquery(c2)|
> +------------------+
> |             false|
> |              NULL|
> +------------------+
> ```
> 
> The result is wrong and it should be false, false

After this PR, the result will be false, false

> 
> `select f from (select false as f, max(c1) from t1 where c1 = c2)` is to build a global aggregate that has no aggregate function. If SQL syntax supports group by no column, then we can simply write `select false as f from t1 group by () where c1 = c2`

After this PR, already is.

```
== Parsed Logical Plan ==
'Project [unresolvedalias(scalar-subquery#245 [], None)]
:  +- 'Project ['f]
:     +- 'SubqueryAlias __auto_generated_subquery_name
:        +- 'Project [false AS f#244, unresolvedalias('max('c1), None)]
:           +- 'Filter ('c1 = 'c2)
:              +- 'UnresolvedRelation [t1], [], false
+- 'UnresolvedRelation [t2], [], false

== Analyzed Logical Plan ==
scalarsubquery(c2): boolean
Project [scalar-subquery#245 [c2#246] AS scalarsubquery(c2)#252]
:  +- Project [f#244]
:     +- SubqueryAlias __auto_generated_subquery_name
:        +- Aggregate [false AS f#244, max(c1#248) AS max(c1)#251]
:           +- Filter (c1#248 = outer(c2#246))
:              +- SubqueryAlias spark_catalog.default.t1
:                 +- View (`spark_catalog`.`default`.`t1`, [c1#248])
:                    +- Project [cast(col1#249 as int) AS c1#248]
:                       +- LocalRelation [col1#249]
+- SubqueryAlias spark_catalog.default.t2
   +- View (`spark_catalog`.`default`.`t2`, [c2#246])
      +- Project [cast(col1#247 as int) AS c2#246]
         +- LocalRelation [col1#247]

== Optimized Logical Plan ==
Project [false AS scalarsubquery(c2)#252]
+- LocalRelation [col1#247]

== Physical Plan ==
*(1) Project [false AS scalarsubquery(c2)#252]
+- *(1) LocalTableScan [col1#247]
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 10]
```diff
@@ -599,10 +599,39 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
...
+            case _ =>
+              false
+          }
+        }
+
         if (resultWithZeroTups.isEmpty) {
           // CASE 1: Subquery guaranteed not to have the COUNT bug because it evaluates to NULL
           // with zero tuples.
           planWithoutCountBug
+        } else if (mayHaveCountBug.getOrElse(false) && resultFoldable &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:628]: can you show me which test will fail if we don't have the changes here?

Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:628]: Test case: 
image
Remove `resultFoldable`: 
image
Should be:
image

Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:628]: > can you show me which test will fail if we don't have the changes here?

if without this, the logic will use `Literal(0)` as result to replace `count(1)`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:628]: are you using the latest Spark version? It works fine if I only fix `l.copy(value = null)`
```
scala> sql("SELECT (SELECT count(1) FROM t2 WHERE t1.c1 = t2.c2) FROM t1").show
+------------------+
|scalarsubquery(c1)|
+------------------+
|                 1|
|                 0|
+------------------+
```

Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:628]: Yes, after I merge the code. The test will passed. But the plan will be
```
== Optimized Logical Plan ==
Project [if (isnull(alwaysTrue#253)) 0 else count(1)#250L AS scalarsubquery(c2)#251L]
+- Join LeftOuter, (c1#247 = c2#245)
   :- Project [col1#246 AS c2#245]
   :  +- LocalRelation [col1#246]
   +- Aggregate [c1#247], [count(1) AS count(1)#250L, c1#247, true AS alwaysTrue#253]
      +- Project [col1#248 AS c1#247]
         +- LocalRelation [col1#248]
```
not
```
== Optimized Logical Plan ==
Project [false AS scalarsubquery(c2)#252]
+- LocalRelation [col1#247]
```

Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:628]: Should I split it in two PR? Or just chagne title and description?

Comment (Hisoka-X) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:628]: There are some change after I create this PR. Seem like `resultFoldable` are useful when I create PR but not now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 11]
```diff
@@ -599,6 +599,7 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
         if (Utils.isTesting) {
           assert(mayHaveCountBug.isDefined)
         }
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:602]: unnecessary change

**Comment by cloud-fan** [comment]
@Hisoka-X I think it's clearer to use my example to explain the bug. `NullPropagation` triggers the bug for `(count(1)) is null`, but the root cause is the wrong handling of literals when dealing with the count bug. The bug can be triggered when the scalar subquery has a global aggregate with a constant, which can be produced by `NullPropagation` or by the user query directly.

File: sql/core/src/test/resources/sql-tests/inputs/subquery/scalar-subquery/scalar-subquery-select.sql [hunk 12]
```diff
@@ -235,3 +235,6 @@ SELECT c, (
     FROM (VALUES (0, 6), (1, 5), (2, 4), (3, 3)) t1(a, b)
     WHERE a + b = c
 ) FROM (VALUES (6)) t2(c);
+
+-- SPARK-43156: scalar subquery with Literal result like `COUNT(1) is null`
+SELECT *, (SELECT count(1) is null FROM t2 WHERE t1.c1 = t2.c1) FROM t1
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply sql/core/src/test/resources/sql-tests/inputs/subquery/scalar-subquery/scalar-subquery-select.sql]: Done

**Review by jchen5 — COMMENTED** [review]
Thanks for the updates, it's much clearer now. Looks good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 13]
```diff
@@ -459,7 +459,7 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
       case alias @ Alias(_: AttributeReference, _) =>
         (alias.exprId, Literal.create(null, alias.dataType))
       case alias @ Alias(l: Literal, _) =>
-        (alias.exprId, l.copy(value = null))
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:462]: @Hisoka-X Can we add a comment here to explain why we want to keep this literal value and how it is related to the COUNT bug?

**Review by allisonwang-db — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.4!

**Comment by cloud-fan** [comment]
It's better to have this fix in 3.4, @Hisoka-X can you open a new PR for 3.4 to backport https://github.com/apache/spark/pull/40946 and this PR together?

**Comment by Hisoka-X** [comment]
> It's better to have this fix in 3.4, @Hisoka-X can you open a new PR for 3.4 to backport #40946 and this PR together?

Please check #40977

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM. Thank you all.
